### PR TITLE
Addon-docs: support vue inline rendering

### PR DESCRIPTION
**Pull request by Aaron-Pool**
Issue: #6706 

## What I did

- Added a parameter to `docs` configuration object called `prepareForInline`
- Added dependency [@egoist/vue-to-react](https://github.com/egoist/vue-to-react) for the inline transform
- Added a vue -> react transformation to the `vue/config.js` file in the `addon-docs` folder
- ~~Removed explicit heights from the MDX `<Story />` blocks in the vue-kitchen-sink example~~ (Put these back, since inline stories will be opt-in for now)
- Added a first pass at docs for inline parameters
- Added an inline example to Vue kitchen sink

## How to test

- Check that the inline story in Vue Kitchen sink behaves as expected.
- Add `inlineStories: true` to the vue kitchen sink config, build it and make sure everything looks the way you expect
- Add `inlineStories: true` to the angular kitchen sink config and ensure the appropriate error is thrown
- Read Documentation additions and edit/critique/remove as you see fit 👍 

## Discussion points:

- This kinda breaks the `centered` add-on in docs mode. I tried to disable it, but since it's implemented as a decorator, I couldn't figure out how to use the `match` parameter the same way that @shilman is using to disable the background addon.
- When an explicit height was set on the the `<Story>` block it seemed to cause some overflow issues. Removing the height made everything work fine, but maybe some `overflow` styles should be added in case the user _wants_ to set an explicit height for some reason? I was assuming that was a cross-framework issue, so I didn't worry about addressing the issue here.

### Why not just use a decorator? Why the new docs config property?
This could totally be my lack of knowledge of the codebase, but I couldn't figure out any way to _guarantee_ that my decorator would be the _last one_ to run. And after the transform to React land runs, the ship has sailed. None of the other decorators can run at that point because we're dealing with a react element now. The most straightforward way, from my perspective, was to add a new config that allows you to specify _how_ a story should be prepared to be inlined. Or how it should be cleanly introduced to the React environment. If any of the core maintainers think a decorator would be cleaner/more consistent and there's a way I can easily guarantee my vue-to-react decorator is the last thing in the decorator pipeline to run, I'd be more than glad to pivot on that approach.

Oh, and last but not least, decorators registered to the vue storybook have a type signature of `DecoratorFunction<VueConstructor>` which, by design, _is not_ what the inline transformer returns.

### Why not `vuera` (or some other library)? It has so many 🌟 's
Vuera is pretty robust, heavy, and handles a bunch of edge cases that I don't think we really care about all that much in our use case. `@egoist/vue-to-react`, on the other hand, is [dead-simple react hook single-file implementation](https://github.com/egoist/vue-to-react/blob/master/src/index.js) that should do everything we need it to do. Egoist is an incredibly talented open source developer who has worked in the Vue community for a long time now, and he wrote this tiny library for our _exact_ use-case, to convert Vue components to be renderable in his custom component [demo tool](https://github.com/egoist/demoboard).

Looking forward to comments and criticism 👍

### CODE REVIEW

**Review by shilman — APPROVED** [review]
This is AMAZING. I want to release this ASAP in 5.3. My one concern is that it's a breaking change, so we should probably make it opt-in in 5.3 and then make it the default in 6.0 (I think?!). At any rate, great work!! 👍 👍 👍

**Comment by shilman** [comment]
PS - looks like some snapshots are failing?

**Comment by Aaron-Pool** [comment]
@Shilman for the "breaking change" are you talking about it messing up with the centered add-on, or it messing up when an explicit height is set? Or something else entirely? I think either of those two can actually be fixed to avoid any breaking changes. It just seemed like that work was outside of the scope of this particular PR.

For the snapshot tests that are failing, all the tests from `yarn test` were passing locally. Is there some other suite of tests I need to be running?

**Comment by Aaron-Pool** [comment]
Wait. Scratch that last. I forgot to update snapshots after I merged with latest.

The only snapshot differences is that, previously the story context had an empty object when no doc options were specified, e.g. - `docs: {}` in the parameter object. Whereas now that field is omitted. Do you think that's problematic?

File: examples/vue-kitchen-sink/__snapshots__/vueshots.test.js.snap [hunk 1]
```diff
@@ -525,69 +346,7 @@ exports[`Storyshots Custom|Decorator for Vue With Data 1`] = `
...
-  "kind": "Custom|Decorator for Vue",
-  "name": "withData",
-  "story": "withData",
-  "customContext": 52,
-  "parameters": {
-    "options": {
-      "hierarchyRootSeparator": {},
-      "hierarchySeparator": {}
-    },
+    "docs": {},
```
Comment (Aaron-Pool) [inline examples/vue-kitchen-sink/__snapshots__/vueshots.test.js.snap]: Changes added an empty docs parameter object here.

File: examples/vue-kitchen-sink/__snapshots__/vueshots.test.js.snap [hunk 2]
```diff
@@ -424,33 +278,9 @@ exports[`Storyshots Button Square 1`] = `
...
 exports[`Storyshots Core|Parameters Passed To Story 1`] = `
 <div>
-  Parameters are {"options":{"hierarchyRootSeparator":{},"hierarchySeparator":{}},"globalParameter":"globalParameter","framework":"vue","chapterParameter":"chapterParameter","storyParameter":"storyParameter","displayName":"passed to story"}
-</div>
-`;
-
-exports[`Storyshots Core|Parameters passed to story 1`] = `
-<div>
-  Parameters are {"options":{"hierarchyRootSeparator":{},"hierarchySeparator":{}},"globalParameter":"globalParameter","framework":"vue","chapterParameter":"chapterParameter","storyParameter":"storyParameter"}
+  Parameters are {"options":{"hierarchyRootSeparator":{},"hierarchySeparator":{}},"docs":{},"globalParameter":"globalParameter","framework":"vue","chapterParameter":"chapterParameter","storyParameter":"storyParameter","displayName":"passed to story"}
```
Comment (Aaron-Pool) [inline examples/vue-kitchen-sink/__snapshots__/vueshots.test.js.snap]: Changes added an empty docs parameter object here.

**Comment by Aaron-Pool** [comment]
Ok. Snapshots updated. If I'm being totally honest, I'm not sure why it keeps completely adding and removing "obsolete" snapshots 🤷‍♂ 

The only _actually changed_ snapshots are [here](https://github.com/storybookjs/storybook/pull/7929/commits/584b4896968fc81efa525e8e23a0ae84e43458d7#r319523638) and [here](https://github.com/storybookjs/storybook/pull/7929/commits/584b4896968fc81efa525e8e23a0ae84e43458d7#r319523530) and I would think this would be an acceptable nominal change? But if not, let me know!

**Comment by shilman** [comment]
@Aaron-Pool the breaking change is that before if the user does zero-config their stories get rendered in an iframe by default, whereas with this change their stories will get rendered inline by default. this is totally diff behavior and i imagine will break some peoples' setups.

Instead, `inferInlineStories` should probably still return `false` for Vue. And user's should use `addParameters({ docs: { inlineStories: true } })` to opt-in to the new behavior. In 6.0, we should add back your change to `inferInlineStories` so that inline becomes the default.

I think....

**Comment by Aaron-Pool** [comment]
@Shilman oh! That totally makes sense. Would you like me to go ahead and make that change in this PR?

Also, it seems like it would be pretty easy to provide a prop option for the doc story block to specify inline true/false on a per story basis? Then people who wanted to go ahead and opt-in and then add `inline=false` to any stories they needed to keep iframe based?

**Comment by shilman** [comment]
@Aaron-Pool Yeah please make the change. There already is an `inline` prop for the Story block. One step ahead of ya, buddy 😘

**Comment by Aaron-Pool** [comment]
Oh... Well. That's embarrassing 😳

I'll make that change ASAP tho. Appreciate the patience 👌

**Comment by Aaron-Pool** [comment]
Updated to be a non-breaking change.

I also took a pass at Docs, since it's an opt-in thing and added an inline/not-inline example to the vue kitchen sink. Feel free to throw away any or all of this if it's not up to par. I won't be offended at all. Just figured I make as well take a crack at it while everything I did was fresh in my mind.

**Comment by shilman** [comment]
@Aaron-Pool The docs additions are great. This is a really solid PR!

I'm wondering if it makes sense to add a little error checking. E.g. if I specify `inlineStories: true` or `inline=true` on the component, there better either be a `prepareForInline` either provided by the preset or by the user. React's could be the identity function. Throwing a nice error with a pointer to the docs will be a lot easier for users to debug. And I bet people will get this wrong.

What do you think?

File: addons/docs/src/frameworks/vue/config.js [hunk 3]
```diff
@@ -1,10 +1,17 @@
 /* eslint-disable import/no-extraneous-dependencies */
+import React from 'react';
+import { render } from 'react-dom';
```
Comment (shilman) [inline addons/docs/src/frameworks/vue/config.js]: Unused? @Aaron-Pool

**Comment by Aaron-Pool** [comment]
@Shilman yep, I can see an invalid inline property combination being a headache for the user. I'll get some error checking in there 👌

File: addons/docs/src/frameworks/vue/config.js (hunk 3 above)
Comment (Aaron-Pool) [reply addons/docs/src/frameworks/vue/config.js]: @shilman I thought React had to be in scope to use JSX? (The render is unused though, I suppose. I think these may have both auto imported when I added JSX to the file?)

**Comment by Aaron-Pool** [comment]
@shilman added inline config validation, as requested

File: addons/docs/src/frameworks/vue/config.js (hunk 3 above)
Comment (shilman) [reply addons/docs/src/frameworks/vue/config.js]: @Aaron-Pool Yeah was only referring to the `render` import

**Comment by shilman** [comment]
@Aaron-Pool This looks amazing. See my comment about the `render` import. Can't wait to merge this as soon as 5.2 goes out!

**Comment by ndelangen** [comment]
@shilman bump

**Comment by Aaron-Pool** [comment]
I still have to make one small change of removing an unused import. @shilman asked me to do it ages ago, and I've just been lazy about it because it wouldn't be merged until 5.2, but I guess we're at that point

**Pull request merged**